File: old-src/shared/Codecs/Transform.lua
local COORD_MULTIPLIER = 5 -- Numbers have accuracy within 0.2
local ORIENTATION_MULTIPLIER = 100 -- Numbers have accuracy within 0.01

local ReplicatedStorage = game:GetService("ReplicatedStorage")
local Components = require(ReplicatedStorage.Shared.Components)

type TransformComponent = Components.TransformComponent

local Transform = {}

Transform.replicationRate = 0.166

function Transform.encode(transformData)
    -- Orientation has range [0, 2pi)
    local orientation = math.floor(ORIENTATION_MULTIPLIER * transformData.orientation + 0.5)
    local x = math.floor(COORD_MULTIPLIER * transformData.position.X + 0.5)
    local z = math.floor(COORD_MULTIPLIER * transformData.position.Z + 0.5)

    return { Vector3int16.new(orientation, x, z) }
end

function Transform.decode(positioningData)
    local orientation = positioningData[1].X / ORIENTATION_MULTIPLIER
    local x = positioningData[1].Y / COORD_MULTIPLIER
    local z = positioningData[1].Z / COORD_MULTIPLIER

    local y = 2 -- calculateHeightOfUnit(x, z)

    local transform: TransformComponent = {
        orientation = orientation,
        position = Vector3.new(x, y, z),
    }
    return transform
end

return Transform

File: src/shared/codecs/transform.ts
const coordMultiplier = 5;
const orientationMultiplier = 100;

type TransformData = {
	position: Vector3;
	orientation: number;
};

type PositioningData = [
	{
		X: number;
		Y: number;
		Z: number;
	},
];

function encode(transformData: TransformData): [Vector3int16] {
	const orientation = math.floor(orientationMultiplier * transformData.orientation + 0.5);
	const x = math.floor(coordMultiplier * transformData.position.X + 0.5);
	const z = math.floor(coordMultiplier * transformData.position.Z + 0.5);

	return [new Vector3int16(orientation, x, z)];
}

function decode(positioningData: PositioningData): TransformData {
	const data = positioningData[0];

	const orientation = data.X / orientationMultiplier;
	const x = data.Y / coordMultiplier;
	const z = data.Z / coordMultiplier;

	const y = 2; // calculateHeightOfUnit(x, z)

	const transform: TransformData = {
		orientation: orientation,
		position: new Vector3(x, y, z),
	};
	return transform;
}

export { encode, decode };
